cpu/stm32l1/periph/pm.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2015 Freie Universität Berlin
+ *               2015 Engineering-Spirit
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_stm32_common
+ * @ingroup     drivers_periph_pm
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the kernels power management interface
+ *
+ * @author      Nick v. IJzendoorn <nijzndoorn@engineering-spirit.nl>
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ * @author      Fabian Nack <nack@inf.fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "cpu.h" // For numbers of ports
+#include "stmclk.h"
+#include "periph/pm.h"
+#include "periph/spi.h"
+#include "periph_conf.h"
+#include "periph/init.h"
+#include "periph/uart.h"
+#include "uart_stdio.h"
+
+#ifdef MODULE_XTIMER
+#include "xtimer.h"
+#endif
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @name Available power modes
+ */
+enum pm_mode {
+    PM_OFF = 0,            /**< MCU is off */
+    PM_POWERDOWN,      /**< MCU is powered down */
+    PM_SLEEP,          /**< MCU in sleep mode */
+};
+
+#ifndef PM_STOP_CONFIG
+/**
+ * @brief Define config flags for stop mode
+ *
+ * Available values can be found in reference manual, PWR section, register CR.
+ */
+#define PM_STOP_CONFIG (PWR_CR_LPDS | PWR_CR_FPDS)
+#endif
+
+/* Variables definitions to LPM */
+static uint32_t pm_gpio_moder[8];
+static uint32_t pm_gpio_pupdr[8];
+static uint16_t pm_gpio_otyper[8];
+static uint16_t pm_gpio_odr[8];
+static uint8_t  pm_usart[UART_NUMOF];
+static uint32_t ahb_gpio_clocks;
+static uint32_t tmpreg;
+
+static uint16_t pm_portmask_system[8] = { 0 };
+static uint16_t pm_portmask_user[8] = { 0 };
+
+static inline GPIO_TypeDef *_port(gpio_t pin)
+{
+    return (GPIO_TypeDef *)(pin & ~(0x0f));
+}
+
+/**
+ * @brief   Extract the port number form the given identifier
+ *
+ * The port number is extracted by looking at bits 10, 11, 12, 13 of the base
+ * register addresses.
+ */
+static inline int _port_num(gpio_t pin)
+{
+    return ((pin >> 10) & 0x0f);
+}
+
+/**
+ * @brief   Extract the pin number from the last 4 bit of the pin identifier
+ */
+static inline int _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+/* We are not using gpio_init as it sets GPIO clock speed to maximum */
+/* We add GPIOs we touched to exclusion mask pm_portmask_system */
+static void pin_set(gpio_t pin, uint8_t value) {
+    GPIO_TypeDef *port = _port(pin);
+    int pin_num = _pin_num(pin);
+    tmpreg = port->MODER;
+    tmpreg &= ~(3 << (2*pin_num));
+    tmpreg |= (1 << (2*pin_num));
+    port->MODER = tmpreg;
+
+    port->PUPDR &= ~(3 << (2*pin_num));
+    port->OTYPER &= ~(1 << pin_num);
+    if (value) {
+        port->ODR |= (1 << pin_num);
+    } else {
```

Comment:
`else {` goes at the next line. See our coding conventions for more information.